### ISSUE TITLE
rustic check aborts when some special files are in repository

### ISSUE THREAD

**Issue by borkd** [rustic-rs/rustic#39]
`rustic check` fails early on a subset of my known to be good restic repos (v1)

Restic says repo is clean, but due for prune
```
[...]
Jun 30 22:50:04 00:00:29 00:00:00 1373 additional files were found in the repo, which likely contain duplicate data.
Jun 30 22:50:04 00:00:29 00:00:00 You can run `restic prune` to correct this.
Jun 30 22:50:04 00:00:29 00:00:00 check snapshots, trees and blobs
Jun 30 22:59:37 00:10:02 00:09:33 [9:32] 100.00%  831 / 831 snapshots
Jun 30 22:59:37 00:10:02 00:00:00
Jun 30 22:59:37 00:10:02 00:00:00 no errors were found
```

rustic check on it:
```
checking snapshots and trees...
 - reading snapshots...
[00:00:00] ████████████████████████████████████████        831/831                                                                        - checking trees...
[00:00:04] ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░          1/831                                                                       thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: missing field `device` at line 1 column 291', src/blob/tree.rs:203:33
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

with full backtrace:
```
[...]
checking snapshots and trees...
 - reading snapshots...
[00:00:00] ████████████████████████████████████████        831/831 

 - checking trees...
[00:00:05] ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░          1/831                                                                       thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: missing field `device` at line 1 column 291', src/blob/tree.rs:203:33
stack backtrace:
   0:     0x5693454187cd - std::backtrace_rs::backtrace::libunwind::trace::hb729d9642bb971eb
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x5693454187cd - std::backtrace_rs::backtrace::trace_unsynchronized::h373bb774579df5c7
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5693454187cd - std::sys_common::backtrace::_print_fmt::hfbd4e92d240c89bb
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x5693454187cd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h8f618991fbf64972
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x56934543be9c - core::fmt::write::hc69b5b640d88cce8
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/fmt/mod.rs:1196:17
   5:     0x569345412631 - std::io::Write::write_fmt::h3403cef06a24a303
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/io/mod.rs:1654:15
   6:     0x569345419fe5 - std::sys_common::backtrace::_print::h368f27cdedea0e52
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x569345419fe5 - std::sys_common::backtrace::print::ha105c9cf5a64cd17
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x569345419fe5 - std::panicking::default_hook::{{closure}}::h48ed2c3707d5e20e
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:295:22
   9:     0x569345419c59 - std::panicking::default_hook::h8744fc5cea5e3110
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:314:9
  10:     0x56934541a5b8 - std::panicking::rust_panic_with_hook::hc82286af2030e925
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:698:17
  11:     0x56934541a467 - std::panicking::begin_panic_handler::{{closure}}::h1c15057c2f09081f
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:588:13
  12:     0x569345418c84 - std::sys_common::backtrace::__rust_end_short_backtrace::h65de906a5330f8da
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x56934541a199 - rust_begin_unwind
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
  14:     0x569344e72d03 - core::panicking::panic_fmt::h741cfbfc95bc6112
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
  15:     0x569344e72df3 - core::result::unwrap_failed::h995262f85f9c4e2c
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/result.rs:1785:5
  16:     0x569344ea1bdc - futures_util::stream::try_stream::TryStreamExt::try_poll_next_unpin::h9e27e51963d91f34
  17:     0x569344fbaff0 - rustic_rs::commands::execute::{{closure}}::h2d8eac611ebf7679
  18:     0x569344f85f8d - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h3d0482824ccc2c38
  19:     0x569344f5aefd - std::thread::local::LocalKey<T>::with::h8c4b741e4425a44f
  20:     0x569344eb296a - tokio::park::thread::CachedParkThread::block_on::ha865f4ee00444c1b
  21:     0x569344edfe3d - tokio::runtime::thread_pool::ThreadPool::block_on::haf56b5342415c866
  22:     0x569344efa77b - tokio::runtime::Runtime::block_on::h8333755947d7437e
  23:     0x569344ee34ca - rustic_rs::main::h5ee36d2599f6e96b
  24:     0x569344ea1c63 - std::sys_common::backtrace::__rust_begin_short_backtrace::hcedd0aca8c9a1910
  25:     0x569344eceaed - std::rt::lang_start::{{closure}}::h311ae3624bbdb5f7
  26:     0x56934540b7be - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hf833e7144973d4be
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:280:13
  27:     0x56934540b7be - std::panicking::try::do_call::h79761d203bfb6b46
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:492:40
  28:     0x56934540b7be - std::panicking::try::h0561cbbe1722251d
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:456:19
  29:     0x56934540b7be - std::panic::catch_unwind::hbca347ddd031b141
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panic.rs:137:14
  30:     0x56934540b7be - std::rt::lang_start_internal::{{closure}}::h0492050ad281ec32
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/rt.rs:128:48
  31:     0x56934540b7be - std::panicking::try::do_call::h3ebce69871996bb3
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:492:40
  32:     0x56934540b7be - std::panicking::try::hbed537d20e728475
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:456:19
  33:     0x56934540b7be - std::panic::catch_unwind::h4185e2024c6a5d05
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panic.rs:137:14
  34:     0x56934540b7be - std::rt::lang_start_internal::h1899cfd715ca6829
                               at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/rt.rs:128:20
  35:     0x569344ee35a2 - main
  36:     0x72ec1b0542e1 - __libc_start_main
  37:     0x569344e72faa - _start
  38:                0x0 - <unknown>
```

**Comment by aawsome** [comment]
@borkd Thanks for reporting.
Actually I feel very sorry that I left you to find all the bugs related to special files which, in fact, are not really supported ATM, see #40.

Independent from #40 it is a bug that `check` or `prune` do fail if those files are present in a repo.

**Comment by aawsome** [comment]
I checked the JSON format used in restic and only found one thing: If the device for your special file is 0, then restic will not serialize the `device` item. rustic however expects a `device` item for devices and chardevs. I changed this such that a default of 0 is taken. This should fix this problem. 

However, we need to take special car when fixing #40 that those cases are working correctly during `restore` and the like.

@borkd May I ask you to run another test with #41? Thanks a lot!
